fly/src/syntax.rs

```diff
@@ -10,6 +10,73 @@ use serde::Serialize;
 
 use crate::ouritertools::OurItertools;
 
+/// A Sort represents a collection of values, which can be the built-in boolean
+/// sort or a named sort (coming from a Signature).
+#[allow(missing_docs)]
+#[derive(PartialEq, Eq, Clone, Debug, Hash, Serialize, PartialOrd, Ord)]
+pub enum Sort {
+    Bool,
+    Uninterpreted(String),
+}
+
+impl Sort {
+    /// Smart constructor for uninterpreted sort that takes &str
+    pub fn uninterp(name: &str) -> Self {
+        Sort::Uninterpreted(name.to_string())
+    }
+}
+
+impl From<&str> for Sort {
+    /// This is mostly for the Binder smart constructor, making it possible to
+    /// pass either Sort, &Sort, or &str
+    fn from(value: &str) -> Self {
+        Self::uninterp(value)
+    }
+}
+
+impl From<&Sort> for Sort {
+    /// This is mostly for the Binder smart constructor, making it possible to
+    /// pass either Sort, &Sort, or &str
+    fn from(value: &Self) -> Self {
+        value.clone()
+    }
+}
+
+impl fmt::Display for Sort {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            Sort::Bool => "bool".to_string(),
+            Sort::Uninterpreted(i) => i.to_string(),
+        };
+        write!(f, "{s}")
+    }
+}
+
+/// A binder is a variable name and a sort (used e.g. for a quantifier)
+#[derive(PartialEq, Eq, Clone, Debug, Hash, PartialOrd, Ord)]
+pub struct Binder {
+    /// Bound name
+    pub name: String,
+    /// Sort for this binder
+    pub sort: Sort,
+}
+
+impl Binder {
+    /// Smart constructor for a Binder that takes arguments by reference.
+    ///
+    /// TODO(oded): Tej, I made name &str instead of AsRef<str> and everything
+    ///             seems fine. Is that okay?
```

Comment:
```suggestion
```

Yup, this seems fine. Simpler types are nice because they better inform type inference and produce better error messages.